meps/mep-cross-references.md

```diff
@@ -0,0 +1,397 @@
+---
+title: Cross Reference Simplifications using Markdown Links
+mep:
+  id: <0001 - Add when this MEP becomes Active>
+  created: <2023-01-dd - date MEP is active>
+  authors:
+    - Chris Sewell @chrisjsewell
+    - Rowan Cockett @rowanc1
+    - Franklin Koch @fwkoch
+  status: Draft
+  discussion: https://github.com/executablebooks/myst-enhancement-proposals/issues/9
+---
+
+## Summary
+
+We propose a cross-reference syntax that uses CommonMark links to support all use cases of cross-referencing both internally to a project and externally using intersphinx. The syntax aims to be familiar and work across different rendering platforms. Most internal content can be referenced using a hash-link, `[](#my-id)`, which is the recommended replacement for the multiple role options that can do this in MyST currently (e.g. `` {ref}`my-id` ``, `` {eq}`my-id` ``, `` {numref}`my-id` ``). We also introduce a `myst` scheme/protocol, which allows for rich cross-project referencing and is compatible with intersphinx. For example, a `[](myst:python#zipfile.ZipFile)` link will create a cross-reference to the Python documentation. We provide options for increasing specificity for these links in all cases to deal with duplicate references across pages, domains, and projects.
+
+| Existing Syntax                                   | New Syntax                                 |
+| :------------------------------------------------ | :----------------------------------------- |
+| `[](my-id)`[^legacy_hash]                         | `[](#my-id)`                               |
+| `` {ref}`my-id` ``                                | `[](#my-id)`                               |
+| `` {eq}`my-equation` ``                           | `[](#my-equation)`                         |
+| `` {ref}`Custom Text <my-id>` ``                  | `[Custom Text](#my-id)`                    |
+| `` {ref}`See "{name}" <my-id>` ``                 | `[See "%t"](#my-id)`                       |
+| `` {numref}`Custom Number %s <my-id>` ``          | `[Custom Number %s](#my-id)`               |
+| `` {doc}`my-doc` ``                               | `[](my-doc.md)`                            |
+| `` {doc}`my-doc` ``                               | `[](../examples/my-doc.md)`                |
+| `` {download}`my-doc.zip` ``                      | `[](my-doc.zip)`                           |
+| `` {py:class}`zipfile.ZipFile` ``                 | `[](#zipfile.ZipFile)`                     |
+| `` {external+python:py:class}`zipfile.ZipFile` `` | `[](myst:python?py:class#zipfile.ZipFile)` |
```

Comment:
In https://github.com/executablebooks/MyST-Parser/pull/673 I have changed this slightly (see also https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#cross-project-inventory-links):

1. `inv` instead of `myst`, to denote that it is referencing an `objects.inv` file
2. `python:py:class` instead of `python?py:class`, where e.g. `python:py` would be resolved as `python:py:*` and `python` as `python:*:*`
3. `*` is a wildcard that can match zero or more characters, and `\*` matches the literal `*`

Reply:
The moving of `python?py:class` --> `python:py:class` makes sense to me, as well as the wildcard.

Note about `inv` in comments below.

Reply:
I have removed all mentions of intersphinx now, and am opting to discuss this in a future MEP. This keeps this one smaller and focust on internal project cross references.